prov/gni/test/av.c

```diff
@@ -317,6 +361,422 @@ Test(av_full_table, lookup)
 	lookup_test();
 }
 
+/* Stuff for mulithreaded tests */
+static pthread_barrier_t mtbar;
+
+/* Currently the AV operations are not thread safe, so use this big
+ * fat lock when calling them */
+#define USE_LOCK
+#ifdef USE_LOCK
+static fastlock_t my_big_lock;
+#define init_av_lock() fastlock_init(&my_big_lock)
+#define av_lock() fastlock_acquire(&my_big_lock)
+#define av_unlock() fastlock_release(&my_big_lock)
+#else
+#define init_av_lock()
+#define av_lock()
+#define av_unlock()
+#endif
+
+static void *insert_single(void *data)
+{
+	int ret;
+	int n = (int) ((int *) data)[0];
+	fi_addr_t *addr = ((fi_addr_t **) data)[1];
+
+	ret = pthread_barrier_wait(&mtbar);
+	if ((ret != PTHREAD_BARRIER_SERIAL_THREAD) && (ret != 0)) {
+		pthread_exit((void *) 1UL);
+	}
+
+	av_lock();
+	ret = fi_av_insert(av, (void *) &simple_ep_names[n], 1, addr, 0, NULL);
+	av_unlock();
+
+	if (ret != 1) {
+		pthread_exit((void *) 2UL);
+	}
+
+	pthread_exit((void *) 0UL);
+}
+
+static void *remove_single(void *data)
+{
+	int ret;
+	fi_addr_t *addr = (fi_addr_t *) data;
+
+	ret = pthread_barrier_wait(&mtbar);
+	if ((ret != PTHREAD_BARRIER_SERIAL_THREAD) && (ret != 0)) {
+		pthread_exit((void *) 1UL);
+	}
+
+	av_lock();
+	ret = fi_av_remove(av, addr, 1, 0);
+	av_unlock();
+
+	if (ret != FI_SUCCESS) {
+		pthread_exit((void *) 2UL);
+	}
+
+	pthread_exit((void *) 0UL);
+}
+
+static void *lookup_single(void *data)
+{
+	int ret;
+	struct gnix_ep_name found;
+	size_t addrlen = sizeof(struct gnix_ep_name);
+	fi_addr_t *addr = (fi_addr_t *) data;
+
+	ret = pthread_barrier_wait(&mtbar);
+	if ((ret != PTHREAD_BARRIER_SERIAL_THREAD) && (ret != 0)) {
+		pthread_exit((void *) 1UL);
+	}
+
+	av_lock();
+	ret = fi_av_lookup(av, *addr, &found, &addrlen);
+	av_unlock();
+
+	if (ret != FI_SUCCESS) {
+		pthread_exit((void *) 2UL);
+	}
+
+	pthread_exit((void *) 0UL);
+}
+
+static void simple_mt_test(void)
+{
+	int ret;
+	unsigned long pret;
+	int i, j;
+	fi_addr_t *compare;
+	pthread_t threads[MT_ADDR_COUNT];
+	fi_addr_t addresses[MT_ADDR_COUNT];
+	bool found_addresses[MT_ADDR_COUNT];
+	void *info[MT_ADDR_COUNT][2];
+
+	ret = pthread_barrier_init(&mtbar, NULL, MT_ADDR_COUNT);
+	cr_assert_eq(ret, 0);
+
+	init_av_lock();
+
+	/* insert addresses */
+	for (i = 0; i < MT_ADDR_COUNT; i++) {
+		info[i][0] = (void *) (uint64_t) i;
+		info[i][1] = (void *) &addresses[i];
+		ret = pthread_create(&threads[i], NULL,
+				     insert_single, &info[i]);
+		cr_assert_eq(ret, 0);
+	}
+	for (i = 0; i < MT_ADDR_COUNT; i++) {
+		ret = pthread_join(threads[i], (void **) &pret);
+		cr_assert_eq(ret, 0);
+		cr_assert_eq(pret, 0UL);
+	}
+
+	for (i = 0; i < MT_ADDR_COUNT; i++) {
+		found_addresses[i] = false;
+	}
+
+	/* check address contents */
+	for (i = 0; i < MT_ADDR_COUNT; i++) {
+		if (gnix_av->type == FI_AV_MAP) {
+			for (j = 0; j < MT_ADDR_COUNT; j++) {
+				compare = (fi_addr_t *)
+					&simple_ep_names[j].gnix_addr;
+				if (addresses[i] == *compare) {
```

Comment:
Is the fi_addr_t (in the addresses array created by fi_av_insert) guaranteed to be the same as the gnix address?  If not, can we use this method of verifying the address contents or do we need to translate each gnix_addr to a fi_addr_t before comparing?  I thought the gnix addresses and fabric addresses contained different contents.